### PR TITLE
oidc-signin-tool: replace puppeteer with playwright

### PR DESCRIPTION
**Pull request by ben-polinsky**
Replaces the use of puppeteer with playwright for the oidc-signin-tool:

- No need to mock the response to the callback url, instead grab the url with code and state by waiting for the redirect request.
- Replace puppeteer syntax and clunky waits
- Add option to simulate slow network condition when launching browsers to ensure waits work as expected.

#93

### CODE REVIEW

**Comment by ben-polinsky** [comment]
@CalebGerman @arvindsvenkat I am wondering about "Authing" support. @aruniverse and I determined this is support for China? I could not test these changes with that workflow since our test user has exceeded some limits on their account.

**Comment by ben-polinsky** [comment]
Still need to verify:
- consent screen
- authing support

**Comment by ben-polinsky** [comment]
(Ugh bad branch name... will just leave it)

File: packages/oidc-signin-tool/src/test-integration/basic.test.ts [hunk 1]
```diff
@@ -1,157 +1,170 @@
...
       scope: process.env.IMJS_OIDC_AUTHING_BROWSER_TEST_SCOPES ?? "",
     };
     if (process.env.IMJS_OIDC_AUTHING_BROWSER_TEST_CLIENT_SECRET)
-      authingOidcConfig.clientSecret = process.env.IMJS_OIDC_AUTHING_BROWSER_TEST_CLIENT_SECRET;
+      authingOidcConfig.clientSecret =
+        process.env.IMJS_OIDC_AUTHING_BROWSER_TEST_CLIENT_SECRET;
   });
 
-  it("success AzureAD with valid user", async () => {
+  // reached active user limit
```
Comment (aruniverse) [inline packages/oidc-signin-tool/src/test-integration/basic.test.ts:166]: looks to be skipping the azure tests not authing?

File: packages/oidc-signin-tool/src/test-integration/basic.test.ts [hunk 2]
```diff
@@ -1,157 +1,170 @@
...
     azureAdOidcConfig = {
       authority: process.env.IMJS_OIDC_AZUREAD_BROWSER_TEST_AUTHORITY,
       clientId: process.env.IMJS_OIDC_AZUREAD_BROWSER_TEST_CLIENT_ID ?? "",
-      redirectUri: process.env.IMJS_OIDC_AZUREAD_BROWSER_TEST_REDIRECT_URI ?? "",
+      redirectUri:
+        process.env.IMJS_OIDC_AZUREAD_BROWSER_TEST_REDIRECT_URI ?? "",
       scope: process.env.IMJS_OIDC_AZUREAD_BROWSER_TEST_SCOPES ?? "",
     };
 
     // Authing oidc config
```
Comment (aruniverse) [inline packages/oidc-signin-tool/src/test-integration/basic.test.ts:138]: maybe we should pull out checking these env vars are defined into a fn, and call that fn only when youre about to test that specific authority

File: packages/oidc-signin-tool/package.json [hunk 3]
```diff
@@ -35,14 +35,14 @@
     "@itwin/certa": "^3.3.0",
     "dotenv": "^10.0.0",
     "dotenv-expand": "^5.1.0",
-    "openid-client": "^4.7.4",
-    "puppeteer": "^13.5.2"
+    "openid-client": "^4.7.4"
```
Comment (aruniverse) [inline packages/oidc-signin-tool/package.json:38]: should eventually switch this to use openid-client-ts like the browser auth did

File: packages/oidc-signin-tool/package.json [hunk 4]
```diff
@@ -35,14 +35,14 @@
...
   },
   "devDependencies": {
     "@itwin/build-tools": "^3.3.0",
     "@itwin/core-bentley": "^3.3.0",
     "@itwin/core-common": "^3.3.0",
     "@itwin/eslint-plugin": "^3.3.0",
+    "@playwright/test": "~1.31.2",
```
Comment (aruniverse) [inline packages/oidc-signin-tool/package.json:45]: i believe this needs to stay a direct dep; so when others use this pkg on their end, playwright will be installed on their end. also we use `chromium` from this pkg in the TestBrowserAuthClient, (so even though it says test in the name, its what consumers will actually be using)

File: common/changes/@itwin/oidc-signin-tool/bdp-signin-tool-use-puppeteer-93_2023-03-10-16-55.json [hunk 5]
```diff
@@ -0,0 +1,10 @@
+{
+  "changes": [
+    {
+      "packageName": "@itwin/oidc-signin-tool",
+      "comment": "Move from puppeteer to playwright",
+      "type": "patch"
```
Comment (aruniverse) [inline common/changes/@itwin/oidc-signin-tool/bdp-signin-tool-use-puppeteer-93_2023-03-10-16-55.json:6]: wonder if this a big enough change we use minor?

**Review by aruniverse — COMMENTED** [review]
lgtm; didnt actually test locally tho. famous last words...

**Review by aruniverse — APPROVED** [review]

File: packages/oidc-signin-tool/src/test-integration/basic.test.ts (hunk 2 above)
Comment (ben-polinsky) [reply packages/oidc-signin-tool/src/test-integration/basic.test.ts:138]: Yeah definitely, this is very tough to read.

File: packages/oidc-signin-tool/src/test-integration/basic.test.ts (hunk 1 above)
Comment (ben-polinsky) [reply packages/oidc-signin-tool/src/test-integration/basic.test.ts:166]: Not sure what happened there, thanks for catching it

File: packages/oidc-signin-tool/package.json (hunk 4 above)
Comment (ben-polinsky) [reply packages/oidc-signin-tool/package.json:45]: Yep thanks.

File: common/changes/@itwin/oidc-signin-tool/bdp-signin-tool-use-puppeteer-93_2023-03-10-16-55.json (hunk 5 above)
Comment (ben-polinsky) [reply common/changes/@itwin/oidc-signin-tool/bdp-signin-tool-use-puppeteer-93_2023-03-10-16-55.json:6]: I wavered back and forth tbh

File: packages/oidc-signin-tool/package.json (hunk 3 above)
Comment (aruniverse) [reply packages/oidc-signin-tool/package.json:38]: ah yes, youre right. stupid o**-client pkg name

**Comment by ben-polinsky** [comment]
@aruniverse even though the tests pass here, there is a flaky test on windows I'd like to try to strengthen. I've been testing locally my Mac so hopefully won't be too tough to recreate on a pc.

**Comment by calebmshafer** [comment]
> @calebmshafer @arvindsvenkat I am wondering about "Authing" support. @aruniverse and I determined this is support for China? I could not test these changes with that workflow since our test user has exceeded some limits on their account.

@ben-polinsky no worries, you can go ahead and comment that out or just delete that test.

File: common/changes/@itwin/oidc-signin-tool/bdp-signin-tool-use-puppeteer-93_2023-03-10-16-55.json (hunk 5 above)
Comment (ben-polinsky) [reply common/changes/@itwin/oidc-signin-tool/bdp-signin-tool-use-puppeteer-93_2023-03-10-16-55.json]: Minor okay?

File: packages/oidc-signin-tool/package.json (hunk 4 above)
Comment (ben-polinsky) [reply packages/oidc-signin-tool/package.json]: fixed

File: packages/oidc-signin-tool/src/test-integration/basic.test.ts (hunk 1 above)
Comment (ben-polinsky) [reply packages/oidc-signin-tool/src/test-integration/basic.test.ts:91]: Made this change

File: packages/oidc-signin-tool/src/test-integration/basic.test.ts (hunk 2 above)
Comment (ben-polinsky) [reply packages/oidc-signin-tool/src/test-integration/basic.test.ts]: Done this as well

File: common/changes/@itwin/oidc-signin-tool/bdp-signin-tool-use-puppeteer-93_2023-03-10-16-55.json (hunk 5 above)
Comment (aruniverse) [reply common/changes/@itwin/oidc-signin-tool/bdp-signin-tool-use-puppeteer-93_2023-03-10-16-55.json]: yea lets do it

Comment (ben-polinsky) [reply common/changes/@itwin/oidc-signin-tool/bdp-signin-tool-use-puppeteer-93_2023-03-10-16-55.json]: it's done if you view all commits

**Review by aruniverse — APPROVED** [review]
lgtm!

**Pull request merged**